pkg/cloudprovider/vsphere/nodemanager_test.go

```diff
@@ -267,3 +268,825 @@ func TestReturnIPsFromSpecificFamily(t *testing.T) {
+						IpAddress: []string{
+							"127.0.0.6",
+						},
+					},
+					{
+						Network: "internal_net",
+						IpAddress: []string{
+							"10.10.5.8",
+						},
+					},
+					{
+						Network: "external_net",
+						IpAddress: []string{
+							"172.15.2.3",
+						},
+					},
+				},
+			},
+			expectedIPs: []v1.NodeAddress{
+				{Type: "InternalIP", Address: "10.10.5.8"},
+			},
+		},
+		{
+			testName: "BySubnet_whenOnlyExternalCIDRIsSet_itReturnsOnlyExternalIP",
+			setup: testSetup{
+				ipFamilyPriority: []string{"ipv4"},
+				cpiConfig: &ccfg.CPIConfig{
+					Nodes: ccfg.Nodes{
+						ExternalNetworkSubnetCIDR: "172.15.0.0/16",
+					},
+				},
+				networks: []vimtypes.GuestNicInfo{
+					{
+						Network: "test_k8s_tenant_c123",
+						IpAddress: []string{
+							"127.0.0.6",
+							"20.30.40.50",
+							"10.10.1.22",
+							"10.10.1.23",
+							"172.15.108.10",
+							"172.15.108.11",
+						},
+					},
+				},
+			},
+			expectedIPs: []v1.NodeAddress{
+				{Type: "ExternalIP", Address: "172.15.108.10"},
+			},
+		},
+		{
+			testName: "BySubnet_whenOnlyInternalCIDRIsSet_itReturnsOnlyInternalIP",
+			setup: testSetup{
+				ipFamilyPriority: []string{"ipv4"},
+				cpiConfig: &ccfg.CPIConfig{
+					Nodes: ccfg.Nodes{
+						InternalNetworkSubnetCIDR: "10.10.0.0/16",
+					},
+				},
+				networks: []vimtypes.GuestNicInfo{
+					{
+						Network: "test_k8s_tenant_c123",
+						IpAddress: []string{
+							"127.0.0.6",
+							"20.30.40.50",
+							"10.10.1.22",
+							"10.10.1.23",
+							"172.15.108.10",
+							"172.15.108.11",
+						},
+					},
+				},
+			},
+			expectedIPs: []v1.NodeAddress{
+				{Type: "InternalIP", Address: "10.10.1.22"},
+			},
+		},
+		{
+			// Is this desirable behavior? It is going through the fallback behavior.
+			// The inverse of this case (ByNetwork_whenOnlyInternalNetworkIsSet_itReturnsOnlyInternalIP)
+			// has a different behavior, I would at least expect these two
+			// cases to have similar behavior. Note: this isn't because the
+			// InternalVMNetworkName acts diffferently from the
+			// ExternalVMNetworkName, but because of which NIC it is on.
+			testName: "ByNetwork_whenOnlyExternalNetworkIsSet_itReturnsOnlyExternalIP",
+			setup: testSetup{
+				ipFamilyPriority: []string{"ipv4"},
+				cpiConfig: &ccfg.CPIConfig{
+					Nodes: ccfg.Nodes{
+						ExternalVMNetworkName: "test_k8s_tenant_c123_external",
+					},
+				},
+				networks: []vimtypes.GuestNicInfo{
+					{
+						Network: "test_k8s_tenant_c123_internal",
+						IpAddress: []string{
+							"127.0.0.6",
+							"10.10.1.22",
+							"10.10.1.23",
+						},
+					},
+					{
+						Network: "test_k8s_tenant_c123_external",
+						IpAddress: []string{
+							"127.0.0.7",
+							"172.15.108.10",
+							"172.15.108.11",
+						},
+					},
+				},
+			},
+			expectedIPs: []v1.NodeAddress{
+				{Type: "InternalIP", Address: "10.10.1.22"},
+				{Type: "ExternalIP", Address: "10.10.1.22"},
+			},
```

Comment:
This seems unexpected. The desired IP exists on the second NIC.

Reply:
🤔  after checking the code I believe the expected IP should be "172.15.108.10"

Reply:
@nicolehanjing I agree, this is incorrect behavior. I'd like to keep this PR as merely test coverage (noting this expectation is undesireable), and then follow up with a PR to refactor and correct these problems. The PR will highlight the bug fixes/changes in behavior.